### PR TITLE
feat: add strict mode option

### LINKED ISSUE #46: Dokumentationsanpassungen
Dokumentieren, dass in die `validate` Funktion nur vertrauenswürdige Funktionen übergeben werden dürfen

### LINKED ISSUE #48: validate: Zusätzlicher `options` Parameter mit Flag `strict: boolean`, das dann prüft ob die übergebenen Funktionen bekannte Tests sind.

### PR DESCRIPTION
**Pull request by domachine**
Closes #46, closes #48.

### CODE REVIEW

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please state more clearly that passing an function that is not defined in the lib may result in an (unintended) code execution. The calling function is in such case responsible of checking whether that execution is wanted or not.

File: lib/validate.js [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
+  .concat(Object.values(schema))
+
 /**
  * @param {Array<import('./shared/types').DocumentTest>} tests
  * @param {any} doc
+ * @param {object} [options]
+ * @param {boolean} [options.strict]
  */
-export default async function (tests, doc) {
+export default async function (tests, doc, { strict = true } = {}) {
```
Comment (tschmidtb51) [inline lib/validate.js:19]: Why do we do that as an object? For extensibility? => Great.

However, I would need a closer look at `{ strict = true } = {}`. That looks interesting as it seems to be a double assignment that I can't figure out right now. I should probably look at that with fresh eyes.

Comment (domachine) [reply lib/validate.js:19]: This double equal sign is due to the feature that the strict option is optional as well as the complete options object.

**Comment by domachine** [comment]
> Please state more clearly that passing an function that is not defined in the lib may result in an (unintended) code execution. The calling function is in such case responsible of checking whether that execution is wanted or not.

Do you have a suggestion for a good description?

File: lib/validate.js (hunk 1 above)
Comment (tschmidtb51) [reply lib/validate.js:19]: Thanks

File: lib/validate.js [hunk 2]
```diff
@@ -1,13 +1,30 @@
...
   let isValid = true
   const testResults =
     /** @type {({ name: string } & import('./shared/types').Result)[]} */ ([])
 
   for (const test of tests) {
+    if (strict && !validTests.includes(test)) {
+      throw new Error('Insecure test function detected')
```
Comment (tschmidtb51) [inline lib/validate.js:26]:
```suggestion
      throw new Error('Execution of test functions not defined in the library is prohibited. See https://github.com/secvisogram/csaf-validator-lib#strict-mode for more details.')

File: tests/validate.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        [
+          function () {
+            return {}
+          },
+        ],
+        {}
+      )
+      expect.fail()
+    } catch (/** @type {any} */ e) {
+      expect(e.message).to.equal('Insecure test function detected')
```
Comment (tschmidtb51) [inline tests/validate.js:17]: See above

**Comment by tschmidtb51** [comment]
> > Please state more clearly that passing an function that is not defined in the lib may result in an (unintended) code execution. The calling function is in such case responsible of checking whether that execution is wanted or not.
> 
> Do you have a suggestion for a good description?

I didn't when I wrote the first review - so I just told the concept what I think we need.

File: README.md [hunk 4]
```diff
@@ -200,10 +200,13 @@ export const informativeTest_6_3_11: DocumentTest
 
 #### Module `validate.js`
 
+This function validates the given document against the given tests. The `strict` option (default `true`) throws if an unknown test function was passed.
```
Comment (tschmidtb51) [inline README.md:203]: Maybe also link here to the section strict mode:
```suggestion
This function validates the given document against the given tests. The `strict` option (default `true`) throws an error if an unknown test function was passed. See [Strict Mode](#strict-mode) for details.

File: README.md [hunk 5]
```diff
@@ -216,10 +219,13 @@ export default ValidateFn
 
 #### Module `strip.js`
 
+This function strips empty nodes and nodes with errors. The `strict` option (default `true`) throws if an unknown test function was passed.
```
Comment (tschmidtb51) [inline README.md:222]: Same applies here.

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
I would suggest to add a section "Strict Mode" in the the README. Here is a rough draft. Feel free to adopt.

```
## Strict Mode

In the default setting, the library checks whether the test that should be executed was defined in the library. Otherwise, it throws an error.
To extend the library, that check can be turned of. **In such case, the calling function is responsible of checking whether the test function passed to the `csaf-validator-lib` is benign. Calling arbitrary functions (especially those resulting from user input) may result in a code execution vulnerability. Therefore, the check of the test function to determine whether it is benign MUST be done before calling it.**
To proceed this dangerous path, set `strict = false`.

File: tests/validate.js (hunk 3 above)
Comment (valentindey) [reply tests/validate.js:17]:
```suggestion
      expect(e.message).to.contain('Execution of test functions not defined in the library is prohibited.')
```

**Review by valentindey — DISMISSED** [review]

**Comment by valentindey** [comment]
> I would suggest to add a section "Strict Mode" in the the README. Here is a rough draft. Feel free to adopt.
> 
> ```
> ## Strict Mode
> 
> In the default setting, the library checks whether the test that should be executed was defined in the library. Otherwise, it throws an error.
> To extend the library, that check can be turned of. **In such case, the calling function is responsible of checking whether the test function passed to the `csaf-validator-lib` is benign. Calling arbitrary functions (especially those resulting from user input) may result in a code execution vulnerability. Therefore, the check of the test function to determine whether it is benign MUST be done before calling it.**
> To proceed this dangerous path, set `strict = false`.
> ```

To me that description would be clear. Just make sure to correct the typo "turned _off_* when adding such a section @domachine :slightly_smiling_face:

File: README.md [hunk 6]
```diff
@@ -62,6 +63,14 @@ subtree in your repository. After that you can reference the modules from within
 
 [(back to top)](#bsi-csaf-validator-lib)
 
+## Strict Mode
```
Comment (tschmidtb51) [inline README.md:66]: Maybe, use one level more... (it still kinda belongs to "How to use")
```suggestion
### Strict Mode

File: README.md [hunk 7]
```diff
@@ -62,6 +63,14 @@ subtree in your repository. After that you can reference the modules from within
...
+
+In the default setting, the library checks whether the test that should be executed was defined in the library. Otherwise, it throws an error.
+To extend the library, that check can be turned off. **In such case, the calling function is responsible of checking whether the test function passed to the `csaf-validator-lib` is benign. Calling arbitrary functions (especially those resulting from user input) may result in a code execution vulnerability. Therefore, the check of the test function to determine whether it is benign MUST be done before calling it.**
```
Comment (tschmidtb51) [inline README.md:69]: I try to limit the bold to the important parts.
```suggestion
To extend the library, that check can be turned off. In such case, **the calling function is responsible for checking** whether the test function passed to the `csaf-validator-lib` is benign. **Calling arbitrary** functions (especially those resulting from user input) may result in a **code execution vulnerability**. Therefore, the check of the test function to determine whether it is benign **MUST be done before calling** it.

File: package-lock.json [hunk 8]
```diff
@@ -2688,5 +2689,6 @@
       "integrity": "sha512-rVksvsnNCdJ/ohGc6xgPwyN8eheCxsiLM8mxuE/t/mOVqJewPuO1miLpTHQiRgTKCLexL4MeAFVagts7HmNZ2Q==",
       "dev": true
     }
-  }
+  },
+  "version": "0.1.0"
```
Comment (tschmidtb51) [inline package-lock.json:2693]: Are both of them updated automatically?

File: README.md [hunk 9]
```diff
@@ -200,7 +209,7 @@ export const informativeTest_6_3_11: DocumentTest
 
 #### Module `validate.js`
 
-This function validates the given document against the given tests. The `strict` option (default `true`) throws if an unknown test function was passed.
+This function validates the given document against the given tests. The `strict` option (default `true`) throws an error if an unknown test function was passed. See [Strict Mode](#strict-mode) for details.
```
Comment (tschmidtb51) [inline README.md:212]:
```suggestion
This function validates the given document against the given tests. The `strict` option (default `true`) throws an error if an unknown test function was passed. See [Strict Mode](#strict-mode) for more details.

**Review by tschmidtb51 — COMMENTED** [review]
Only some minor wording and format things - other than that okay for me.

File: package-lock.json (hunk 8 above)
Comment (domachine) [reply package-lock.json:2693]: Yes this was done using `npm version minor`

Comment (tschmidtb51) [reply package-lock.json:2693]: Perfect.

**Review by tschmidtb51 — APPROVED** [review]
LGTM

**Pull request merged**